AnyText/AnyText.Core/Parser.Formatting.cs

```diff
@@ -0,0 +1,152 @@
+        /// </summary>
+        public TextEdit[] Format(
+            ParsePosition? start = null,
+            ParsePosition? end = null,
+            uint tabsize = 4,
+            bool insertSpaces = true,
+            Dictionary<string, object> otherOptions = null,
+            bool trimTrailingWhitespace = false,
+            bool insertFinalNewline = false,
+            bool trimFinalNewlines = false)
+        {
+            var (startPos, endPos) = GetStartAndEndPositions(start, end);
+
+            if (Context.Errors.Any())
+                return new TextEdit[] { };
+
+            var ruleApp = Context.Matcher.GetRuleApplicationsAt(startPos).FirstOrDefault(r => r.Rule.IsLiteral);
+            if (ruleApp == null)
+                return new TextEdit[] { };
+
+            while (!(ruleApp.CurrentPosition <= startPos && ruleApp.CurrentPosition + ruleApp.Length >= endPos))
+            {
+                ruleApp = ruleApp.Parent;
+                if (ruleApp == null)
+                    return new TextEdit[] { };
+            }
+
+            var formattedLines = GetFormattedLines(tabsize, insertSpaces, ruleApp);
+
+            formattedLines = ApplyFormattingOptions(
+                formattedLines, trimTrailingWhitespace, insertFinalNewline, trimFinalNewlines);
+
+            var originalLines = Context.Input.Skip(startPos.Line).Take(endPos.Line - startPos.Line + 1).ToArray();
+
+            // Check if FormattedLines includes chars beyond the range
+            if (!CompareStringArraysByContent(originalLines, formattedLines.ToArray())) return new TextEdit[] { };
+
+            return GenerateTextEdits(startPos, endPos, originalLines, formattedLines);
+        }
+
+        private (ParsePosition startPos, ParsePosition endPos) GetStartAndEndPositions(ParsePosition? start,
+            ParsePosition? end)
+        {
+            start ??= new ParsePosition(0, 0);
+            end ??= GetEndPosition();
+            return (start.Value, end.Value);
+        }
+
+        private bool CompareStringArraysByContent(string[] array1, string[] array2)
+        {
+            var firstString = RemoveWhitespace(string.Join("", array1));
+            var secondString = RemoveWhitespace(string.Join("", array2));
+
+            return firstString.Length == secondString.Length;
+        }
+
+        private string RemoveWhitespace(string input)
+        {
+            return new string(input.Where(c => !char.IsWhiteSpace(c)).ToArray());
+        }
+
+        private ParsePosition GetEndPosition()
+        {
+            var lastLineIndex = _context.Input.Length - 1;
+            var lastLine = _context.Input[lastLineIndex];
+            return new ParsePosition(lastLineIndex, lastLine.Length + 1);
+        }
+
+        private List<string> GetFormattedLines(uint tabsize, bool insertSpaces, RuleApplication ruleApp)
+        {
+            var indentation = insertSpaces ? new string(' ', (int)tabsize) : "\t";
```

Comment:
Ehrlich gesagt würde ich die Logik zum Berechnen der Indentation lieber in den LSP-Server auslagern und nicht im Parser haben wollen.